From 31716a8f2399dfbd10ff370dca03b5b5ecc63ba7 Mon Sep 17 00:00:00 2001
From: Guruswamy Basavaiah <guruswamy.basavaiah@broadcom.com>
Date: Sat, 3 Jan 2026 01:30:07 +0530
Subject: [PATCH] vmgenid: expose vmgenid via sysfs

Make VM generation ID available as a hex string in /sys/kernel/vmgenid.
Using this ID userspace can detect if VM was cloned or modified.

Signed-off-by: Alexey Makhalov <alexey.makhalov@broadcom.com>
[Guru: Ported to 6.12 kernel from 6.1. 6.12 kernel vmgenid_add
function now modified to handle device tree.
Modified the patch accordingly.]
Signed-off-by: Guruswamy Basavaiah <guruswamy.basavaiah@broadcom.com>
---
 drivers/virt/vmgenid.c | 39 ++++++++++++++++++++++++++++++++++++++-
 1 file changed, 38 insertions(+), 1 deletion(-)

diff --git a/drivers/virt/vmgenid.c b/drivers/virt/vmgenid.c
index 66135eac3abf..4192e4fed02e 100644
--- a/drivers/virt/vmgenid.c
+++ b/drivers/virt/vmgenid.c
@@ -21,15 +21,37 @@ enum { VMGENID_SIZE = 16 };
 struct vmgenid_state {
 	u8 *next_id;
 	u8 this_id[VMGENID_SIZE];
+	spinlock_t lock; /* to protect this_id */
+	struct kobj_attribute kobj_attr;
 };
 
+static ssize_t vmgenid_show(struct kobject *kobj, struct kobj_attribute *attr,
+			    char *page)
+{
+	struct vmgenid_state *state = container_of(attr, struct vmgenid_state, kobj_attr);
+	u8 id[VMGENID_SIZE];
+	unsigned long flags;
+
+	/* To avoid reading partially modified ID */
+	spin_lock_irqsave(&state->lock, flags);
+	memcpy(id, state->this_id, sizeof(id));
+	spin_unlock_irqrestore(&state->lock, flags);
+
+	return sysfs_emit(page, "%*phN\n", VMGENID_SIZE, id);
+}
+
 static void vmgenid_notify(struct device *device)
 {
 	struct vmgenid_state *state = device->driver_data;
 	u8 old_id[VMGENID_SIZE];
+	unsigned long flags;
 
 	memcpy(old_id, state->this_id, sizeof(old_id));
+
+	spin_lock_irqsave(&state->lock, flags);
 	memcpy(state->this_id, state->next_id, sizeof(state->this_id));
+	spin_unlock_irqrestore(&state->lock, flags);
+
 	if (!memcmp(old_id, state->this_id, sizeof(old_id)))
 		return;
 	add_vmfork_randomness(state->this_id, sizeof(state->this_id));
@@ -136,7 +158,7 @@ static int vmgenid_add(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
 	struct vmgenid_state *state;
-	int ret;
+	int rc, ret;
 
 	state = devm_kmalloc(dev, sizeof(*state), GFP_KERNEL);
 	if (!state)
@@ -149,6 +171,21 @@ static int vmgenid_add(struct platform_device *pdev)
 
 	if (ret < 0)
 		devm_kfree(dev, state);
+	else {
+		spin_lock_init(&state->lock);
+
+		state->kobj_attr.attr.name = "vmgenid";
+		state->kobj_attr.attr.mode = 0444;
+		state->kobj_attr.show = vmgenid_show;
+		rc = sysfs_add_file_to_group(kernel_kobj, &state->kobj_attr.attr, NULL);
+		if (rc) {
+			/*
+			 * Throw an error message and go forward. Continue using
+			 * vmgenid device without sysfs entry.
+			 */
+			dev_err(dev, "Error adding sysfs file [%d]\n", rc);
+		}
+	}
 	return ret;
 }
 
-- 
2.43.0

